BPF Host Routing: do not skip fib lookups

aspsk opened the pull request:
See the commit message.

<!-- Description of change -->

Fixes: #27343

```release-note
Do not skip FIB lookup when running in BPF Host Routing when Endpoint Routes enabled
```

lmb reviewed (approved):
config change seems fine to me. No idea whether ENABLE_SKIP_FIB is great or not, but I'm all for reducing the number of code paths we have. Unless we have a strong indication that it's hurtful I'd say go for it!

aspsk committed:
BPF Host Routing: do not skip fib lookups on EKS

When cilium is running in the BPF Host Routing mode and DirectRoutingDevice is
set and cilium is configured to only use a single device, then we have skipped
the FIB lookup for packets leaving the pod and forwarded them directly to the
aforementioned DirectRoutingDevice network interface.

This doesn't work in case of native routing when other, non-cilium-managed,
interfaces are present on a node. In such case we, obviously never send traffic
to such interfaces, even for a reply, for example, when the DirectRoutingDevice
interface is eth0 and another eth1 interface is present, then the following
happens:

One known case when this happens on practice is EKS. It can happen that a node
has multiple interfaces, e.g., eth0, eth1, and packets from some pods (to the
VPC private range) are routed through eth1 instead of eth0. For example,

    # ip rule
    20:     from all to 192.168.91.2 lookup main
    100:    from all lookup local
    111:    from 192.168.86.135 to 192.168.0.0/16 lookup 10
    111:    from 192.168.80.149 to 192.168.0.0/16 lookup 11
    32766:  from all lookup main
    32767:  from all lookup default

In this [simplified] output packets from pod 192.168.86.135 to 192.168.0.0/16
(VPC private network) should be routed through the eth0 (table 10 contains a
default route through eth0), and similar packets from pod 192.168.80.149 should
be routed through eth1. This breaks with BPF host routing because, as described
above, all packets go to eth0.

In order to fix this we want to disable the ENABLE_SKIP_FIB macro in BPF code.
This macro is a big optimization, so in this patch we only try to disable it
for configs which are expected to be broken by the change in v1.14.0-snapshot.2
(which allowed to enable endpoint routes and host routing simultaneously). So,
we enforce the FIB lookup when endpoint routes are enabled.

In general, for such setups users should be using `devices=eth+` and
`enableRuntimeDeviceDetection=true`. In future we should make this behaviour
default, and once done, to revert this patch.

Fixes: #27343

Signed-off-by: Anton Protopopov <aspsk@isovalent.com>

lmb commented on pkg/datapath/linux/config/config.go:
```diff
@@ -1073,6 +1074,11 @@ func (h *HeaderfileWriter) WriteEndpointConfig(w io.Writer, e datapath.EndpointC
 	return h.writeTemplateConfig(fw, e)
 }
 
+func enforceFibLookup(e datapath.EndpointConfiguration) bool {
```
Could you put this on the type of `option.Config` instead? Then we wouldn't add another dependency on the global variable.

julianwiedmann reviewed (approved):
Looks good, thank you! But could you please update the commit msg & release notes? :pray: 

@margamanterola maybe you want to confirm that this fixes the issue from https://github.com/cilium/cilium/issues/27343 ?

lmb reviewed (approved):
Thanks :)

derailed commented on pkg/option/config.go:
```diff
@@ -4391,3 +4391,13 @@ func parseBPFMapEventConfigs(confs BPFEventBufferConfigs, confMap map[string]str
+	//
+	// We want to enforce FIB lookup if EndpointRoutes are enabled, because
+	// this was a config dependency change which caused different behaviour
+	// since v1.14.0-snapshot.2. We will remove this hack later, once we
```
Wondering if we can tag these somehow so we can `grep` for code debt in the future?

derailed reviewed (approved):
@aspsk Nice work!
